File: cast-extractor/src/main/java/com/libreshockwave/tools/preview/TextPreview.java
package com.libreshockwave.tools.preview;

import com.libreshockwave.DirectorFile;
import com.libreshockwave.cast.MemberType;
import com.libreshockwave.chunks.TextChunk;
import com.libreshockwave.tools.model.CastMemberInfo;
import com.libreshockwave.tools.scanning.MemberResolver;

/**
 * Generates text member preview content.
 */
public class TextPreview {

    /**
     * Formats text member content for display.
     */
    public String format(DirectorFile dirFile, CastMemberInfo memberInfo) {
        StringBuilder sb = new StringBuilder();
        String typeName = memberInfo.memberType() == MemberType.BUTTON ? "BUTTON" : "TEXT";
        sb.append("=== ").append(typeName).append(": ").append(memberInfo.name()).append(" ===\n\n");
        sb.append("Member ID: ").append(memberInfo.memberNum()).append("\n\n");

        // Find the text chunk for this member
        TextChunk textChunk = MemberResolver.findTextForMember(dirFile, memberInfo.member());

        if (textChunk != null) {
            String text = textChunk.text();
            // Normalize line endings: \r\n -> \n, \r -> \n
            text = text.replace("\r\n", "\n").replace("\r", "\n");

            sb.append("--- Text Content ---\n");
            sb.append(text);
            sb.append("\n\n");

            // Show formatting info if present
            if (!textChunk.runs().isEmpty()) {
                sb.append("--- Formatting Runs ---\n");
                for (TextChunk.TextRun run : textChunk.runs()) {
                    sb.append(String.format("  Offset %d: Font #%d, Size %d, Style 0x%02X\n",
                            run.startOffset(), run.fontId(), run.fontSize(), run.fontStyle()));
                }
            }
        } else {
            sb.append("[Text data not found]\n");
        }

        return sb.toString();
    }
}

File: cpp/src/editor/preview/TextPreview.cpp
#include "libreshockwave/editor/preview/TextPreview.hpp"

#include <cstddef>
#include <iomanip>
#include <sstream>
#include <string>

#include "libreshockwave/cast/MemberType.hpp"
#include "libreshockwave/chunks/TextChunk.hpp"
#include "libreshockwave/editor/preview/PreviewFormatUtils.hpp"
#include "libreshockwave/editor/scanning/MemberResolver.hpp"

namespace libreshockwave::editor::preview {
namespace {

std::string normalizeTextLineEndings(std::string value) {
    std::string normalized;
    normalized.reserve(value.size());
    for (std::size_t index = 0; index < value.size(); ++index) {
        if (value[index] == '\r') {
            normalized.push_back('\n');
            if (index + 1 < value.size() && value[index + 1] == '\n') {
                ++index;
            }
        } else {
            normalized.push_back(value[index]);
        }
    }
    return normalized;
}

} // namespace

std::string TextPreview::format(DirectorFile& dirFile, const model::CastMemberInfo& memberInfo) const {
    std::string out;
    const std::string typeName = memberInfo.memberType == cast::MemberType::Button ? "BUTTON" : "TEXT";
    PreviewFormatUtils::appendMemberHeader(out, typeName, memberInfo, true);

    const auto text = scanning::MemberResolver::findTextForMember(dirFile, memberInfo.member);
    if (text == nullptr) {
        out += "[Text data not found]\n";
        return out;
    }

    out += "--- Text Content ---\n";
    out += normalizeTextLineEndings(text->text());
    out += "\n\n";

    if (!text->runs().empty()) {
        out += "--- Formatting Runs ---\n";
        std::ostringstream row;
        for (const auto& run : text->runs()) {
            row.str("");
            row.clear();
            row << "  Offset " << run.startOffset << ": Font #" << run.fontId << ", Size " << run.fontSize
                << ", Style 0x" << std::uppercase << std::hex << std::setfill('0') << std::setw(2)
                << run.fontStyle << std::dec << std::setfill(' ') << "\n";
            out += row.str();
        }
    }

    return out;
}

} // namespace libreshockwave::editor::preview
